sherpa/astro/data.py

```diff
@@ -1761,27 +1761,87 @@ def get_specresp(self, filter=False):
-    # they should be grouped if the data are grouped.  There are
-    # certain contexts (e.g., plotting) where we will retrieve the
-    # energy bins, and later filter the data; but filtering
-    # is automatically followed by grouping.  Grouping the data
-    # twice is an error.
     def _get_ebins(self, response_id=None, group=True):
+        """Return the low and high edges of the independent axis.
+
+        This method is badly named as it will return values in either
+        channel or energy units, depending on the units setting and
+        the associated response information. It also returns the
+        approximation of the mapping from channel space to energy -
+        that is the E_MIN and E_MAX columns from the RMF EBOUNDS block
+        rather than from the ENERG_LO and ENERG_HI columns from the
+        MATRIX block.
+
+        Parameters
+        ----------
+        response_id : int or None, optional
+            The response to use when units are not "channel". The
+            default is to use the default response identifier.
+        group : bool, optional
+            Should the current grouping setting be applied. This is
+            only used if the "grouped" attribute is set.
+
+        Returns
+        -------
+        lo, hi : ndarray
+            The low and high edges of each bin, in either channels or
+            keV: energy is used if the units setting is not energy and
+            a response (RMF or ARF) is loaded, otherwise channels are
+            used. If the group flag is set and the data set is grouped
+            then it uses the grouping settings, otherwise the data is
+            for each channel. No filtering is applied.
+
+        See Also
+        --------
+        _get_indep
+
+        Examples
+        --------
+
+        >>> pha.ungroup()
+        >>> pha.units = 'channel'
+        >>> clo, chi = pha._get_ebins()
+        >>> (clo == pha.channel).all()
+        True
+        >>> (chi == clo + 1).all()
+        True
+
+        >>> pha.units = 'energy'
+        >>> elo, ehi = pha._get_ebins()
+        >>> elo.size == pha.channel.size
+        True
+        >>> elo[0:5]
+        array([0.00146, 0.0146 , 0.0292 , 0.0438 , 0.0584 ])
+        >>> (elo[1:] == ehi[:-1]).all()
+        True
+
+        >>> pha.group()
+        >>> glo, ghi = pha._get_ebins()
+        >>> glo[0:5]
+        array([0.00146   , 0.2482    , 0.3066    , 0.46720001, 0.56940001])
+
+        Note that the returned units are energy even if units is set
+        to "wavelength":
+
+        >>> pha.units = 'wave'
+        >>> wlo, whi = pha._get_ebins()
+        >>> (wlo == glo).all()
+
+        """
         group = bool_cast(group)
 
         if self.units == 'channel':
             elo = self.channel
             ehi = self.channel + 1
+        elif (self.bin_lo is not None) and (self.bin_hi is not None):
+            elo = self.bin_lo
+            ehi = self.bin_hi
+            if (elo[0] > elo[-1]) and (ehi[0] > ehi[-1]):
```

Comment:
I wonder if we run into trouble with XMM/RGS here, which I've seen reverse-energy sorted (I think. Maybe I'm just confused. It's the end of a long holiday weekend.). I know that this is just the old code moved up by a few lines, so it's probably fine, but I'm not sure I can follow the logic in my head to make sure.
I guess the XMM case will end up in the else clause. But why are we guaranteed that `bin_lo` or `bin_hi` are `None`, so we know that we can get down to the next part of the logic?

Reply:
The code used to be

    if a: ... do a things
    else:
        set up arf, rmf variables
        if b: ... do b things and do not use arf or rmf variables
        elif c: ... do c things
        elif d: ... do d things
        else: ... do boring things

to

    if a: ... do a things
    elif b: ... do b things
    else:
        setup arf, rmf variables
        if c: ... do c things
        elif d: ... do things 
        else: ... do boring t hings

The important thing is that the `arf` and `rmf` variables are not used in the `a` or `b` checks (they are in `c` and `d`). We could move the `arf, rmf = ..` line to before this check and then this becomes a straight `if/elif/...` chain, which might be a bit easier to scan, but not sure it's worth it. For reference the `get_response` method (which returns `arf` and `rmf`) can be safely moved before this section (i.e. it's not going to fall over if condition a is True).